prov/util/src/util_mr_cache.c

```diff
@@ -318,6 +318,46 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
 	return ret;
 }
 
+int ofi_mr_cache_reg(struct ofi_mr_cache *cache, const struct fi_mr_attr *attr,
+		     struct ofi_mr_entry **entry)
+{
+	int ret;
+
+	assert(attr->iov_count == 1);
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "reg %p (len: %" PRIu64 ")\n",
+	       attr->mr_iov->iov_base, attr->mr_iov->iov_len);
+
+	pthread_mutex_lock(&cache->monitor->lock);
+	*entry = ofi_buf_alloc(cache->entry_pool);
+	if (*entry) {
+		cache->uncached_cnt++;
+		cache->uncached_size += attr->mr_iov->iov_len;
+	} else {
+		ret = -FI_ENOMEM;
+		goto unlock;
+	}
+	pthread_mutex_unlock(&cache->monitor->lock);
+
+	(*entry)->info.iov = *attr->mr_iov;
+	(*entry)->use_cnt = 1;
+	(*entry)->cached = 0;
+
+	ret = cache->add_region(cache, *entry);
+	if (ret)
+		goto buf_free;
+
+	return 0;
+
+buf_free:
+	pthread_mutex_lock(&cache->monitor->lock);
+	ofi_buf_free(*entry);
+	cache->uncached_cnt--;
+	cache->uncached_size -= attr->mr_iov->iov_len;
+unlock:
+	pthread_mutex_unlock(&cache->monitor->lock);
+	return ret;
+}
+
```

Comment:
Why do we need this call? Why not make providers directly register instead of calling this?

Reply:
When the cache is used, all registrations are associated with a struct ofi_mr_entry.  That structure is owned by and comes from the cache.  The fid_mr returned by the registration call is contained within this structure.

If we don't go through the cache, then we need to update the registration and deregistration calls to figure out how the fid_mr was allocated (either via the cache or some other structure allocated by the provider), so we can call the correct unregister and free calls.